### PR TITLE
[DRAFT] SR-IOV enhancement

### PR DESCRIPTION
**Pull request by tlehman**
This is a draft, putting it here for visibility. Currently there are two main features here:

1. SR-IOV on the local Harvester cluster and
2. SR-IOV on a virtualized guest cluster using PCI-Passthrough. This will enable customers to work exclusively in the guest cluster and centralize all their configuration there, using PCI-Passthrough to get direct access to their network devices.

**Problem:**
Adding SR-IOV support to Harvester to allow for faster I/O on virtualized workloads. 

**Solution:**
Add sriov-network-operator and custom UI / Go code to integrate it into Harvester.

**Related Issue:**
#992 

**Test plan:**
TODO

### CODE REVIEW

File: enhancements/20220614-sriov-configuration.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# SR-IOV Configuration
```
Comment (guangbochen) [inline enhancements/20220614-sriov-configuration.md:1]: Can we please follow the template from  https://github.com/harvester/harvester/blob/master/enhancements/YYYYMMDD-template.md and add those missing sections.

File: enhancements/20220614-sriov-configuration.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+SR-IOV stands for Single-Root I/O Virtualization. It provides a way for virtual machines to directly access and 
+even share physical network interfaces. SR-IOV drivers create multiple VFs (Virtual Functions). 
+
+![diagram explaining SR-IOV](20220614-sriov-configuration/VFs.png)
+
+VFs are exposed by the kernel as a named NIC. For example, `eth1` is a NIC, and there are 4 VFs, 
+so the OS and the VMs both see `eth1v1`, `eth1v2`, `eth1v3` and `eth1v4`. 
+The PF (Physical Function) is represented by `eth1`, and the VFs are the four `eth1vN` interfaces above.
+
+# Guest cluster mode
```
Comment (guangbochen) [inline enhancements/20220614-sriov-configuration.md:12]: I think SR-IOV support for VMs should be the primary case and then the guest k8s cluster. So the user stories should include:
- Allow users to add an SR-IOV network interface to the VM.
- Support VM live migration connected with an SR-IOV network interface.
- (Experimental) Guest cluster SR-IOV support with leveraging the PCI passthrough.

File: enhancements/20220614-sriov-configuration.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+## Requirements
+1. SR-IOV nav item (under Advanced) is only visible when `sriov-networking-enabled` setting to `true`
+2. Clicking on the SR-IOV nav item will show a form to edit the fields for these two CRDs:
+   2a. [SriovNetwork](https://docs.openshift.com/container-platform/4.6/networking/hardware_networks/configuring-sriov-net-attach.html) and
+   2b. [SriovNetworkNodePolicy](https://docs.openshift.com/container-platform/4.6/networking/hardware_networks/configuring-sriov-device.html**
+3. When the SriovNetworkNodePolicy is created, the spec.resourceName should be the default value for the SriovNetwork's spec.resourceName
+4. the VLAN number in the SriovNetwork must be unique from all the other VLAN numbers currently
+5. nodes UI needs checkbox for 'feature.node.kubernetes.io/network-sriov.capable=true' label
+
+## Technical Details
```
Comment (guangbochen) [inline enhancements/20220614-sriov-configuration.md:90]: Would be better to add the following YAML examples so both UI and backend can know the primary changes. 
- VM example that uses the SR-IOV network.
- Multus NAD example of the SR-IOV network configs.

File: enhancements/20220614-sriov-configuration.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+To make use of the operator, we have a fork of the operator source code here. 
+Repo: https://github.com/harvester/sriov-network-operator
+
+### Installation
+When the `sriov-network-enabled` setting is changed to `true`, the settings controller will
+install the `sriov-network-operator` [Helm chart](https://github.com/harvester/sriov-network-operator/tree/master/deployment/sriov-network-operator).
```
Comment (guangbochen) [inline enhancements/20220614-sriov-configuration.md:96]: We may need a little bit more comparison between using the `sriov-network-operator` vs `vanilla sriov chart` like https://github.com/rancher/charts/pull/1211#pullrequestreview-668492770

**Review by guangbochen — COMMENTED** [review]
Additionally, we should keep the HEP PR with documentation only and leave the code for another PR, thanks.
Note: you can add [skip ci](https://docs.drone.io/pipeline/skipping/) for the doc-related PR.

File: enhancements/20220614-sriov-configuration.md (hunk 2 above)
Comment (tlehman) [reply enhancements/20220614-sriov-configuration.md:12]: I found this [device plugin](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#:~:text=Works%20within%20virtual%20deployments%20of%20Kubernetes%20that%20do%20not%20have%20virtualized%2Diommu%20support%20(VFIO%20No%2DIOMMU%20support)) that says it works with virtual deployments of kubernetes, so I will try this out in the guest cluster mode

**Pull request closed without merging**